### PR TITLE
lua support for tls module

### PR DESCRIPTION
**Pull request by MXWXZ**
Add lua setsockopt support for tls module.

### CODE REVIEW

File: Kbuild [hunk 1]
```diff
@@ -0,0 +1,16 @@
+LUNATIK := lunatik/lua
+
+subdir-ccflags-y := -I${PWD}/${LUNATIK} \
+	-Wall \
+	-D_KERNEL \
+        -D_MODULE \
```
Comment (tammela) [inline Kbuild:6]: Mixed indentation.
Please fix this.

File: Kconfig [hunk 2]
```diff
@@ -1,28 +0,0 @@
-# SPDX-License-Identifier: GPL-2.0-only
```
Comment (tammela) [inline Kconfig]: Why did you remove this file?

File: Kbuild [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+	-D'CHAR_BIT=(8)' \
+	-D'MIN=min' \
+	-D'MAX=max' \
+	-D'UCHAR_MAX=(255)' \
+	-D'UINT64_MAX=((u64)~0ULL)'
+
+obj-y += lunatik/
+obj-m += tls.o
+tls-y := tls_main.o tls_sw.o
+tls-m += tls_device.o tls_device_fallback.o
```
Comment (tammela) [inline Kbuild:16]: This doesn't make much sense.

Why not?
```
obj-$(CONFIG_KTLS) := tls.o tls_main.o tls_sw.o tls_device.o tls_device_fallback.o
```

File: test.c [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+int main() {
+    int sock = socket(AF_INET, SOCK_STREAM, 0);
+    if (sock == -1) {
+        perror("socket creation");
+        exit(EXIT_FAILURE);
+    }
+    struct sockaddr_in sa;
+    memset(&sa, 0, sizeof(sa));
+    sa.sin_family = AF_INET;
+    sa.sin_addr.s_addr = inet_addr("103.118.42.79");
```
Comment (tammela) [inline test.c:25]: While testing we should prefer a controlled environment. This includes de server side as well.
Moving on you should setup a local server with TLS support.

File: test.c [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    if (connect(sock, (struct sockaddr*)&sa, socklen)) {
+        perror("socket connect");
+        exit(EXIT_FAILURE);
+    }
+
+    if (setsockopt(sock, SOL_TCP, TCP_ULP, "tls", sizeof("tls")) == -1) {
+        perror("tls init");
+        exit(EXIT_FAILURE);
+    }
+    char* script="print(\"hello world\")";
```
Comment (tammela) [inline test.c:37]: You can use `'` for strings in Lua.

File: tls_main.c [hunk 6]
```diff
@@ -563,6 +567,40 @@ static int do_tls_setsockopt_conf(struct sock *sk, char __user *optval,
...
+#define TLS_LUA 99
+static DEFINE_MUTEX(mtx);
+static lua_State* L;
+#define raise_err(msg) pr_warn("[lua] %s - %s\n", __func__, msg);
+
+static int do_tls_lua(char __user *optval, unsigned int optlen)
+{
+    int rc = 0;
+    char* file;
+    file = kmalloc(optlen, GFP_ATOMIC);
```
Comment (tammela) [inline tls_main.c:579]: `GFP_ATOMIC` is not needed here in this context.
You can use `GFP_KERNEL`.

File: tls_main.c [hunk 7]
```diff
@@ -563,6 +567,40 @@ static int do_tls_setsockopt_conf(struct sock *sk, char __user *optval,
 	return rc;
 }
 
+#define TLS_LUA 99
+static DEFINE_MUTEX(mtx);
+static lua_State* L;
+#define raise_err(msg) pr_warn("[lua] %s - %s\n", __func__, msg);
+
+static int do_tls_lua(char __user *optval, unsigned int optlen)
```
Comment (tammela) [inline tls_main.c:575]: Code style doesn't follow already existing code.
This is bad. You should always follow the existing code's style.

File: tls_main.c [hunk 8]
```diff
@@ -908,13 +949,21 @@ static int __init tls_register(void)
 	tls_device_init();
 	tcp_register_ulp(&tcp_tls_ulp_ops);
 
+    L = luaL_newstate();
```
Comment (tammela) [inline tls_main.c:952]: Not ideal.
This will share one Lua state between all processes that use TLS.
The mutex access will pay a huge price in a multi-threaded web server.
Also, this would share the scripts between the processes.
What if there are two scripts that define two different functions but with the same name?

File: test.c (hunk 4 above)
Comment (MXWXZ) [reply test.c:25]: Yes this is just a test
In future it will be removed.

File: Kconfig (hunk 2 above)
Comment (MXWXZ) [reply Kconfig]: We move tls module out of linux kernel source so we dont have cmd like `make menuconfig`. For simplification I moved the config to Makefile directly.

Comment (tammela) [reply Kconfig]: Hmmm, that's an assumption that doesn't always hold. For instance a user could clone this repository in-tree and integrate to the in-tree Kbuild system.
Let's re-introduce this file.

Comment (MXWXZ) [reply Kconfig]: reverted

File: tls_main.c [hunk 9]
```diff
@@ -568,17 +568,17 @@ static int do_tls_setsockopt_conf(struct sock *sk, char __user *optval,
 }
 
 #define TLS_LUA 99
-static DEFINE_MUTEX(mtx);
-static lua_State* L;
-#define raise_err(msg) pr_warn("[lua] %s - %s\n", __func__, msg);
+#define lua_error(msg) pr_warn("[lua] %s - %s\n", __func__, msg);
 
-static int do_tls_lua(char __user *optval, unsigned int optlen)
+static int do_tls_execlua(char __user *optval, unsigned int optlen)
```
[the remark this answers is not in the document]
Comment (MXWXZ) [reply tls_main.c:573]: will fix in next PR

File: tls_main.c [hunk 10]
```diff
@@ -568,17 +568,17 @@ static int do_tls_setsockopt_conf(struct sock *sk, char __user *optval,
...
 {
     int rc = 0;
     char* file;
-    file = kmalloc(optlen, GFP_ATOMIC);
+    lua_State* state;
```
[the remark this answers is not in the document]
Comment (MXWXZ) [reply tls_main.c]: will fix in next PR

[the remark this answers is not in the document]
Comment (tammela) [reply tls_main.c]: No need to post pone this minor fix.
Please fix it here.

File: tls_main.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (tammela) [reply tls_main.c:573]: No need to post pone this minor fix.
Please fix it here.

[the remark this answers is not in the document]
Comment (MXWXZ) [reply tls_main.c]: done

File: tls_main.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (MXWXZ) [reply tls_main.c]: done

File: tls_main.c [hunk 11]
```diff
@@ -587,16 +587,21 @@ static int do_tls_lua(char __user *optval, unsigned int optlen)
         goto end;
     }
 
-    mutex_lock(&mtx);
-    if (luaL_dostring(L, file)) {
-        raise_err(lua_tostring(L, -1));
-        mutex_unlock(&mtx);
+    state = luaL_newstate();
```
[the remark this answers is not in the document]
Comment (MXWXZ) [reply tls_main.c]: In next PR I merge it with recv script in a struct to persist the state
